clash-lib/src/Clash/Driver/BrokenGhcs.hs

```diff
@@ -0,0 +1,156 @@
+
+module Clash.Driver.BrokenGhcs where
+
+import Data.Maybe (listToMaybe)
+import Data.Version (Version(Version, versionBranch))
+import GHC.Platform (OS(..))
+
+#if __GLASGOW_HASKELL__ > 810
+import System.Info (fullCompilerVersion)
+#endif
+
+import qualified Clash.Util.Interpolate as I
+import qualified System.Info
+
+#if __GLASGOW_HASKELL__ <= 810
+fullCompilerVersion :: Version
+fullCompilerVersion = System.Info.compilerVersion
+#endif
+
+-- | Current OS. Currently only recognizes Linux, Windows, and macOS.
+os :: OS
+os = case System.Info.os of
+  "darwin" -> OSDarwin
+  "linux" -> OSLinux
+  "mingw32" -> OSMinGW32
+  _ -> OSUnknown
+
+-- | What OS GHC is broken on (or all)
+data BrokenOn = All | SomeOs OS
+
+data GhcVersion
+  = Ghc Int Int Int
+  -- ^ @A.B.C@, for example @9.10.2@
+  deriving (Eq, Ord)
+
+data GhcRange = GhcRange
+  { from :: GhcVersion
+  -- ^ Start of range, inclusive
+  , to :: GhcVersion
+  -- ^ End of range, exclusive
+  }
+
+-- | Check if a 'GhcVersion' is within a 'GhcRange'
+ghcInRange :: GhcVersion -> GhcRange -> Bool
+ghcInRange ghc GhcRange{from, to} = from <= ghc && ghc < to
+
+-- | Construct a range of all GHC versions matching a major version
+ghcMajor :: Int -> Int -> GhcRange
+ghcMajor major minor = GhcRange{from=Ghc major minor 0, to=Ghc (major + 1) 0 0}
+
+data Why = Why
+  { what :: String
+    -- ^ What is broken
+  , solution :: String
+    -- ^ What can be done to work around or solve the issue
+  , issue :: String
+    -- ^ Link to issue
+  , brokenOn :: [(BrokenOn, GhcRange)]
+    -- ^ What operation systems are affected
+  }
+
+-- | Get current GHC version expressed as a triple. It probably does something
+-- non-sensible on unreleased GHCs.
+ghcVersion :: GhcVersion
+ghcVersion = Ghc major minor patch
+ where
+  (major, minor, patch) =
+    case fullCompilerVersion of
+      Version{versionBranch} ->
+        case versionBranch of
+          [] -> (0, 0, 1)
+          [a] -> (a, 0, 1)
+          [a, b] -> (a, b, 1)
+          [a, b, c] -> (a, b, c)
+          (a:b:c:_) -> (a, b, c)
+
+-- | Pretty print 'Why' into an error message
+whyPp :: Why -> String
+whyPp Why{what, solution, issue}= [I.i|
+  Clash has known issues on #{major}.#{minor}.#{patch} on your current
+  OS. While not completely preventing the compiler from working, we recommend
+  switching to another GHC version. Symptoms:
+
+    #{what}
+
+  Consider the following work around or solution:
+
+    #{solution}
+
+  More information can be found at:
+
+    #{issue}
+
+  If you want to ignore this message, pass the following flag to Clash:
+
+    -fclash-ignore-broken-ghcs
+  |]
+ where
+  Ghc major minor patch = ghcVersion
+
+-- | Which GHCs are broken and why
+brokenGhcs :: [Why]
+brokenGhcs = [brokenClashCores, brokenTypeErrors, slowStarts]
+ where
+  brokenClashCores = Why
+    { what = "GHC fails to compile clash-cores"
+    , solution = "Upgrade to GHC 9.4 or downgrade to GHC 8.10"
+    , issue = "<no link>"
```

Comment:
Given that `clash-cores` is unreleased, this might not tell a user that much. The user might very well conclude that there is some functionality that is not available to them, but they might not know what it is. Interestingly, if I [ask my search engine](https://www.startpage.com/sp/search?query=clash-cores&cat=web&pl=opensearch&language=english) what `clash-cores` means, the results are not very helpful.

So I would prefer if we would elaborate a bit more what we are talking about.

Of course, one way we could to that is to create an issue with a description on GitHub, and link to that.

Reply:
This?

Reply:
The error message:
> GHC fails to compile clash-cores

Without explaining to the user what on Cat's green earth clash-cores is in the first place, users might not know what you're talking about and whether they're affected.

So _this_ = clash-cores.

Reply:
Where dit the entry go? I no longer see it.

Reply:
Oh, I get it, instead of explaining what clash-cores is, you just describe that without it, the test suite will not run, without mentioning it by name.

Reply:
Yes, exactly